File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/subscribe/internal/playback/MediaFrameObserver.kt
package com.swmansion.moqkit.subscribe.internal.playback

import uniffi.moq.MoqFrame

internal enum class MediaFrameKind {
    AUDIO,
    VIDEO,
}

internal interface MediaFrameObserver {
    fun onMediaFrame(frame: MoqFrame, kind: MediaFrameKind)

    fun onFrameDiscontinuity(kind: MediaFrameKind, gapUs: Long)
}

internal class CompositeMediaFrameObserver(
    private val observers: List<MediaFrameObserver>,
) : MediaFrameObserver {
    override fun onMediaFrame(frame: MoqFrame, kind: MediaFrameKind) {
        observers.forEach { it.onMediaFrame(frame, kind) }
    }

    override fun onFrameDiscontinuity(kind: MediaFrameKind, gapUs: Long) {
        observers.forEach { it.onFrameDiscontinuity(kind, gapUs) }
    }
}

File: ios/Sources/MoQKit/Subscribe/internal/playback/MediaFrameObserver.swift
import Foundation

enum MediaFrameKind: Sendable {
    case audio
    case video
}

extension MediaFrameKind {
    var eventName: String {
        switch self {
        case .audio:
            return "audio"
        case .video:
            return "video"
        }
    }
}

protocol MediaFrameObserver: Sendable {
    func onMediaTrackStarted(kind: MediaFrameKind)
    func onMediaFrame(kind: MediaFrameKind, frame: MediaFrame)
    func onMediaDiscontinuity(kind: MediaFrameKind, gapUs: UInt64)
}

final class CompositeMediaFrameObserver: MediaFrameObserver {
    private let observers: [any MediaFrameObserver]

    init(_ observers: [any MediaFrameObserver]) {
        self.observers = observers
    }

    func onMediaTrackStarted(kind: MediaFrameKind) {
        for observer in observers {
            observer.onMediaTrackStarted(kind: kind)
        }
    }

    func onMediaFrame(kind: MediaFrameKind, frame: MediaFrame) {
        for observer in observers {
            observer.onMediaFrame(kind: kind, frame: frame)
        }
    }

    func onMediaDiscontinuity(kind: MediaFrameKind, gapUs: UInt64) {
        for observer in observers {
            observer.onMediaDiscontinuity(kind: kind, gapUs: gapUs)
        }
    }
}
